packages/flutter/lib/src/rendering/editable.dart

```diff
@@ -1029,6 +1064,26 @@ class RenderEditable extends RenderBox {
     return null;
   }
 
+  /// Clamps a local coordinate within the viewport.
+  ///
+  /// This is used for example to prevent the text selection handles from
+  /// displaying outside the viewport while fading out.
+  Offset clampToViewport(Offset point) {
+    switch (_viewportAxis) {
+      case Axis.horizontal:
+        final double dx = point.dx.clamp(0.0, _viewportExtent);
+        if (dx != point.dx)
+          return Offset(point.dx.clamp(0.0, _viewportExtent), point.dy);
+        break;
+      case Axis.vertical:
+        final double dy = point.dy.clamp(0.0, _viewportExtent);
+        if (dy != point.dy)
+          return Offset(point.dy, point.dy.clamp(0.0, _viewportExtent));
+        break;
+    }
+    return point;
+  }
```

Comment:
you should test this logic works.
for example, have a test that:
1. selects a word, verifies that opacity is 0.0, waits a second, verifies that opacity is 1.0.
2. scrolls the text field, keeping the handle in the text field, verifies that opacity is 1.0 and in the right place.
3. scrolls the text field, move the handle out. verify that opacity is 1.0 and in the right place (not outside field)
4. waits a tiny bit, verifies that opacity is 0.5 and still in the same place
5. scroll the handle back in, check that opacity is 0.5 but it's no longer stuck on the edge
6. etc

Reply:
I've updated the current test to do check the positioning and animations.